### ISSUE TITLE
Footnote can break to next line on its own when custom font is set

### ISSUE THREAD

**Issue by emilyyyylime** [typst/typst#3129]
### Description

I don't even know what could be causing this but no matter how much I messed with the `width` I could not reproduce this without setting the `font` field of `text`.
Minimal reproducible example:
```typ
#block(width: 4.5cm)[
  #set text(font: ())
  #set par(justify: true)
  Reproduce with English text.#footnote[hi]
]
```
Renders as:
![Rendered using Typst-Renderer on the official Typst Discord: https://discord.com/channels/1054443721975922748/1056574658771681361/1192172744658407504](https://github.com/typst/typst/assets/40892795/e648a888-7c09-4cf3-93df-afb88ed75a23)

### Typst version

- [X] I am using the latest version of Typst

**Comment by ntjess** [comment]
A very hacky workaround in the meantime:
```typst
#let builtin-footnote = footnote
#let footnote(..args) = {
  // Nonbreakable space glues footnote to text
  sym.space.nobreak
  // "box" prevents paragraph break
  // "pad" removes the whitespace added by "nobreak"
  box(pad(left: -0.17em, builtin-footnote(..args)))
}

#block(width: 4.5cm)[
  #set text(font: ())
  #set par(justify: true)
  Reproduce with English text.#footnote[hi]
]
```

**Comment by laurmaedje** [comment]
So, turns out that the issue is the following:

- If possible, a superscript is converted into a specific superscript codepoint
- This is only possible if the active font supports that
- If `set text(font: ())` is set, that's not possible
- Special superscript codepoints have different Unicode linebreaking behaviour

The solution would be to somehow override the linebreak behaviour of non-typographic superscripts (that are just achieved via baseline shift + downscaling).

**Comment by emilyyyylime** [comment]
This is a bit of an edge case, but it's interesting here that we choose to 'fake' superscripts the current font doesn't have 'real' superscripts, even when the current font also doesn't support the faked characters... If someone has some non-latin font configured as their main I would consider this unexpected behaviour

**Comment by laurmaedje** [comment]
> even when the current font also doesn't support the faked characters

What do you mean by this? The "faked" characters are just normal numbers. If the font doesn't have normal numbers, there is not much we can do.

**Comment by emilyyyylime** [comment]
What I mean is, even when creating the 'fake' superscripts as above, aren't we forced to fall back to a different font than the user's choice (`fomt: ()`, meaning the user requests no font to be used, but Typst uses a builtin fallback)

**Comment by emilyyyylime** [comment]
As I understand it, the logic in my `()` example goes like so:
Typst checks if the current font supports the `¹` character; as the current font list contains no fonts, the current font can't support that character; we fallback to `super[1]`. In `super`, Typst typesets the character `1`, once again having to fall back to a built-in font. \
But, the font we've fallen back to for the character `1` _might as well_ contain the character `¹`, so it is in fact rather odd that we didn't use typographic superscripts even though the chosen font does support them

**Comment by laurmaedje** [comment]
Correct. Because the check in the super show rule is rather naive. Aside from the fallback difference, it also couldn't look through show rules for instance, if there were any.

**Comment by orangenschalen** [comment]
The error seems to be fixed as of typst 0.11.0 (2bf9f95d). It no longer breaks between the word and its footnote number with the given code. The word and the superscript number is either completely on the first or completely on a new line

**Comment by laurmaedje** [comment]
Interesting. I can't remember anything change that would fix the root cause, so I wonder what changed this case.

**Comment by YDX-2147483647** [comment]
> A very hacky workaround https://github.com/typst/typst/issues/3129#issuecomment-1939425357

A cleaner workaround:

```typst
#show footnote: it => sym.wj + it
```

Ref:
- [§23.2.1 Line and Word Breaking – Unicode 17.0.0](https://www.unicode.org/versions/Unicode17.0.0/core-spec/chapter-23/#G15334) explains the difference between U+00A0 NO-BREAK SPACE and U+2060 WORD JOINER.
- [WJ: Word Joiner (Non-tailorable) – UAX #14: Unicode Line Breaking Algorithm](https://www.unicode.org/reports/tr14/#WJ) lists other available characters.

<img width="200" alt="Image" src="https://github.com/user-attachments/assets/1dc4aba9-3c17-4fee-aca9-bd326e9ecf22" />

<details>
<summary>Full code</summary>

```typst
#set page(height: auto, width: auto, margin: 1em)
// Using Han characters is easier to predict widths
#set text(lang: "zh", font: "Noto Serif CJK SC")
// Setting a wider numbering makes the problem more visible
#set footnote(numbering: "①")
// Justification is irrelevant to breaking opportunities, but I put it here in case someone wants to test.
#set par(justify: true)


#let test = block.with(width: 6em, stroke: (x: green))

= Before
#test[
  国国：国国国#footnote[😀]
]
= After
#test[
  #show footnote: it => sym.wj + it
  国国：国国国#footnote[😀]
]
```

</details>
